cherab/tools/raytransfer/emitters.py

```diff
@@ -0,0 +1,283 @@
+        The ray tranfer matrix will be calculated only for those cells for which mask is True.
+        This parameter is ignored if `voxel_map` is provided, defaults to `mask=None` (all cells are included).
+    :param raysect.optical.material.VolumeIntegrator integrator: Volume integrator,
+        defaults to `integrator=RayTransferRPhiZIntegrator(step=0.1 * min(dr, dz))`
+    """
+
+    def __init__(self, nr, nz, dr, dz, rmin=0, nphi=0, dphi=1., period=360., voxel_map=None, mask=None, integrator=None):
+        integrator = integrator or RayTransferRPhiZIntegrator(0.1 * min(dr, dz))
+        super().__init__(integrator)
+        if nphi and not dphi * period:
+            raise ValueError('period and dphi must be non-zero for 3D grid')
+        self._rmin = rmin
+        self._period = period
+        self._dr = dr
+        self._dphi = dphi
+        self._dz = dz
+        self._nr = nr
+        self._nphi = nphi
+        self._nz = nz
+        self._map = None
+        self._bins = None
+        if voxel_map is None:
+            self.mask = mask
+        else:
+            self.voxel_map = voxel_map
+
+    def _map_from_mask(self, mask):
+        if mask is not None:
+            if self._nphi and mask.shape != (self._nr, self._nphi, self._nz):
+                raise ValueError('Mask array must be of shape: %d, %d, %d' % (self._nr, self._nphi, self._nz))
+            if (not self._nphi) and mask.shape != (self._nr, self._nz):
+                raise ValueError('Mask array must be of shape: %d, %d' % (self._nr, self._nz))
+            if mask.dtype != np.bool:
+                raise ValueError('Mask array must be of numpy.bool type')
+        else:
+            mask = np.ones((self._nr, self._nphi, self._nz), dtype=np.bool) if self._nphi else np.ones((self._nr, self._nz), dtype=np.bool)
+        voxel_map = -1 * np.ones(mask.shape, dtype=int)
+        voxel_map[mask] = np.arange(mask.sum(), dtype=int)
+
+        return voxel_map
+
+    @property
+    def bins(self):
+        return self._bins
+
+    @property
+    def voxel_map(self):
+        return self._map
+
+    @voxel_map.setter
+    def voxel_map(self, value):
+        if self._nphi and value.shape != (self._nr, self._nphi, self._nz):
+            raise ValueError('Voxel_map array must be of shape: %d, %d, %d' % (self._nr, self._nphi, self._nz))
+        if (not self._nphi) and value.shape != (self._nr, self._nz):
+            raise ValueError('Voxel_map array must be of shape: %d, %d' % (self._nr, self._nz))
+        if value.dtype != np.int:
+            raise ValueError('Voxel_map array must be of numpy.int type')
+        self._map = value
+        self._bins = self._map.max() + 1
+
+    @property
+    def mask(self):
+        return self._map > -1
+
+    @mask.setter
+    def mask(self, value):
+        self._map = self._map_from_mask(value)
+        self._bins = self._map.max() + 1
+
+
+class RayTransferXYZEmitter(InhomogeneousVolumeEmitter):

```

Comment:
I kept an access to protected attributes only in the methods of the same or derived class.

Reply:
I got it. Even derived classes should not have access to the protected attributes of the parent classes, because if the parent class is cythonized later, the derived class may lose access to protected attributes of the parent class. It's fixed now.